crates/turbo-tasks/src/native_function.rs

```diff
@@ -1,18 +1,98 @@
-use std::{fmt::Debug, hash::Hash};
+use std::{fmt::Debug, hash::Hash, pin::Pin};
 
+use anyhow::{Context, Result};
+use futures::Future;
+use serde::{Deserialize, Serialize};
 use tracing::Span;
 
 use crate::{
     self as turbo_tasks,
+    magic_any::{MagicAny, MagicAnyDeserializeSeed, MagicAnySerializeSeed},
     registry::register_function,
     task::{
-        function::{IntoTaskFnWithThis, NativeTaskFn},
+        function::{IntoTaskFnWithThis, NativeTaskFuture},
         IntoTaskFn, TaskFn,
     },
-    util::SharedError,
-    ConcreteTaskInput, RawVc,
+    RawVc, TaskInput,
 };
 
+type ResolveFunctor =
+    for<'a> fn(
+        &'a dyn MagicAny,
+    ) -> Pin<Box<dyn Future<Output = Result<Box<dyn MagicAny>>> + Send + 'a>>;
+
+type IsResolvedFunctor = fn(&dyn MagicAny) -> bool;
+
+pub struct ArgMeta {
+    serializer: MagicAnySerializeSeed,
+    deserializer: MagicAnyDeserializeSeed,
+    is_resolved: IsResolvedFunctor,
+    resolve: ResolveFunctor,
```

Comment:
This is a clever way to avoid propagating the type argument to the struct without dynamic dispatch.